### PR TITLE
Add sum functions

### PR DESCRIPTION
**Pull request by planetis-m**
Alternative to #16004, addresses most review comments. If it's going to be deprecated, I can adopt the package, either works for me. I am the original author.

### CODE REVIEW

File: lib/std/sums.nim [hunk 1]
```diff
@@ -76,3 +103,61 @@ func sumPairs*[T](x: openArray[T]): T =
...
+      let hi = x + y
+      let lo = y - (hi - x)
+      if lo != 0:
+        result[i] = lo
+        inc(i)
+      x = hi
+    setLen(result, i + 1)
+    result[i] = x
+
+func sumShewchuck*[T: SomeFloat](x: openArray[T]): T =
```
Comment (kaushalmodi) [inline lib/std/sums.nim]: Should this be `sumShewchuk` instead of  `sumShewchuck`?

Comment (planetis-m) [inline lib/std/sums.nim]:
```suggestion
func sumShewchuk*[T: SomeFloat](x: openArray[T]): T =
```

File: tests/stdlib/tsums.nim [hunk 2]
```diff
@@ -5,18 +5,24 @@ var epsilon = 1.0
 while 1.0 + epsilon != 1.0:
   epsilon /= 2.0
 let data = @[1.0, epsilon, -epsilon]
-doAssert sumKbn(data) == 1.0
+assert sumKbn(data) == 1.0
+assert sumKbk(data) == 1.0
 # doAssert sumPairs(data) != 1.0 # known to fail in 64 bits
-doAssert (1.0 + epsilon) - epsilon != 1.0
+assert sumShewchuck(data) == 1.0
```
Comment (planetis-m) [inline tests/stdlib/tsums.nim]:
```suggestion
assert sumShewchuk(data) == 1.0
```

File: tests/stdlib/tsums.nim [hunk 3]
```diff
@@ -5,18 +5,24 @@ var epsilon = 1.0
...
 
 var tc1: seq[float]
 for n in 1 .. 1000:
   tc1.add 1.0 / n.float
-doAssert sumKbn(tc1) == 7.485470860550345
-doAssert sumPairs(tc1) == 7.485470860550345
+assert sumKbn(tc1) == 7.485470860550345
+assert sumKbk(tc1) == 7.485470860550345
+assert sumPairs(tc1) == 7.485470860550345
+assert sumShewchuck(tc1) == 7.485470860550345
```
Comment (planetis-m) [inline tests/stdlib/tsums.nim]:
```suggestion
assert sumShewchuk(tc1) == 7.485470860550345
```

File: tests/stdlib/tsums.nim [hunk 4]
```diff
@@ -5,18 +5,24 @@ var epsilon = 1.0
...
 
 var tc2: seq[float]
 for n in 1 .. 1000:
   tc2.add pow(-1.0, n.float) / n.float
-doAssert sumKbn(tc2) == -0.6926474305598203
-doAssert sumPairs(tc2) == -0.6926474305598204
+assert sumKbn(tc2) == -0.6926474305598203
+assert sumKbk(tc2) == -0.6926474305598203
+assert sumPairs(tc2) == -0.6926474305598204
+assert sumShewchuck(tc2) == -0.6926474305598203
```
Comment (planetis-m) [inline tests/stdlib/tsums.nim]:
```suggestion
assert sumShewchuk(tc2) == -0.6926474305598203
```

File: lib/std/sums.nim [hunk 5]
```diff
@@ -23,15 +22,16 @@ runnableExamples:
 
   let result = first + small * n.float
 
-  doAssert abs(sum(data) - result) > 0.3
-  doAssert sumKbn(data) ~= result
-  doAssert sumPairs(data) ~= result
+  assert abs(sum(data) - result) > 0.3
+  assert sumKbn(data) =~ result
+  assert sumKbk(data) =~ result
+  assert sumPairs(data) =~ result
```
Comment (planetis-m) [inline lib/std/sums.nim:30]:
```suggestion
assert sumShewchuk(data) =~ result

```

File: lib/std/sums.nim [hunk 6]
```diff
@@ -23,15 +22,17 @@ runnableExamples:
...
+assert sumShewchuk(data) =~ result
```
Comment (planetis-m) [inline lib/std/sums.nim]:
```suggestion
  assert sumShewchuk(data) =~ result
```

File: tests/stdlib/tsums.nim [hunk 7]
```diff
@@ -5,18 +5,24 @@ var epsilon = 1.0
 while 1.0 + epsilon != 1.0:
   epsilon /= 2.0
 let data = @[1.0, epsilon, -epsilon]
-doAssert sumKbn(data) == 1.0
+assert sumKbn(data) == 1.0
+assert sumKbk(data) == 1.0
 # doAssert sumPairs(data) != 1.0 # known to fail in 64 bits
```
Comment (planetis-m) [inline tests/stdlib/tsums.nim]:
```suggestion
# assert sumPairs(data) != 1.0 # known to fail in 64 bits
```

**Comment by konsumlamm** [comment]
While I like the idea of having different accurate summation functions, I don't think they should be in the standard library (and I'm not sure why this module was added in the first place), but rather a nimble package.

The only real use of `std/sums` I could find (after a quick GitHub search) was in a test, where it only seems to be used to test passing as `openArray` (https://github.com/bpr/vla/blob/master/tests/tests.nim). So, similar to https://github.com/nim-lang/Nim/pull/18395#issuecomment-872967286, I'm in favor deprecating `std/sums`, as it's only very rarely used (definitely not enough to justify having it in the standard library).

**Comment by planetis-m** [comment]
Alright, will do that instead.

**Pull request closed without merging**